Add validation to ConfigMap patch test

BobyMCbobs opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Adds validation to e2e test
```
should patch ConfigMap successfully
```

**Which issue(s) this PR fixes**:
Fixes #78832

**Special notes for your reviewer**:
Part of Umbrella Issue #78748
This test is related to #78924

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/area test
@kubernetes/sig-apps-pr-reviews
@kubernetes/sig-architecture-pr-reviews

BobyMCbobs commented:
/priority backlog

spiffxp commented:
/ok-to-test

>Promotes the following E2E test to Conformance:

please edit your PR description, this adds a new test, doesn't actually promote

/cc @oomichi @spiffxp
/uncc @krmayankk

devaips commented:
/retest

BobyMCbobs committed:
Add ConfigMap patch e2e test with validation step

Test originally written by draveness for this PR:
https://github.com/kubernetes/kubernetes/pull/78924

This commit adds a validation step to the test.

spiffxp commented:
/test pull-kubernetes-integration

timothysc commented on test/e2e/common/configmap.go:
```diff
@@ -134,6 +135,27 @@ var _ = ginkgo.Describe("[sig-node] ConfigMap", func() {
 		framework.ExpectError(err, "created configMap %q with empty key in namespace %q", configMap.Name, f.Namespace.Name)
 	})
+
+	ginkgo.It("should patch ConfigMap successfully", func() {
```
this should be using the framework on this call line vs. ginkgo. 

This applies to everything here.

spiffxp replied:
we don't have a `framework.It`; we do have a `framework.ConformanceIt` but explicitly disallow new tests from being added with ConformanceIt, they must be promoted into it

spiffxp replied:
we don't have a `framework.It`; we do have a `framework.ConformanceIt` but explicitly disallow new tests from being added with ConformanceIt, they must be promoted into it

timothysc replied:
Sorry retract, I'm brain fry, comment below still applies though.

timothysc commented on test/e2e/common/configmap.go:
```diff
@@ -134,6 +135,27 @@ var _ = ginkgo.Describe("[sig-node] ConfigMap", func() {
 	})
+
+	ginkgo.It("should patch ConfigMap successfully", func() {
+		name := "configmap-test-" + string(uuid.NewUUID())
```
Why do this vs. using the utility below?

spiffxp reviewed (changes requested):
one nit

BobyMCbobs committed:
Update function for data inequality verification

BobyMCbobs commented on test/e2e/common/configmap.go:
```diff
@@ -134,6 +135,27 @@ var _ = ginkgo.Describe("[sig-node] ConfigMap", func() {
+		configMapFromUpdate, err := f.ClientSet.CoreV1().ConfigMaps(f.Namespace.Name).Get(name, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+		ginkgo.By(fmt.Sprintf("Verifying update of configMap %v/%v", f.Namespace.Name, configMap.Name))
+		gomega.Expect(configMapFromUpdate.Data).NotTo(gomega.Equal(configMapOriginalState.Data))
```
That's a useful function! Thank you.
I've updated the test to use `framework.ExpectNotEqual` in b10743236ef

spiffxp commented:
/retest
/approve

oomichi commented on test/e2e/common/configmap.go:
```diff
@@ -134,6 +134,27 @@ var _ = ginkgo.Describe("[sig-node] ConfigMap", func() {
+		configMapFromUpdate, err := f.ClientSet.CoreV1().ConfigMaps(f.Namespace.Name).Get(name, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+		ginkgo.By(fmt.Sprintf("Verifying update of configMap %v/%v", f.Namespace.Name, configMap.Name))
+		framework.ExpectNotEqual(configMapFromUpdate.Data, configMapOriginalState.Data)
```
Why don't we check the `Get()` data is the same as `Update()`d data(`"data": "value"`) here?
I mean it is nice to add `framework.ExpectEqual(configMapFromUpdate.Data, configMap.Data)`

BobyMCbobs replied:
configMap.Data's value is changed, so it doesn't make sense to compare against it.

[test/e2e/common/configmap.go: L146-148](https://github.com/ii/kubernetes/blob/b10743236ef08ab8cf4b3cf5ac4ea9911db453fa/test/e2e/common/configmap.go#L146)
```golang
 		configMap.Data = map[string]string{
			"data": "value",
		}
```

oomichi replied:
I am not sure I can follow what you are saying here.
In this test, the config map is updated with
```
 		configMap.Data = map[string]string{
			"data": "value",
		}
```
So the config map date which is gotten at line 153 should be the same as the above value, isn't it?

BobyMCbobs replied:
Perhaps the purpose of the variables is being confused.
`configMap` is the original configMap value.   
`configMapOriginalState` has the value of `configMap` before it's changed.  
`configMapFromUpdate` has the value of `configMap` after it's updated, pulling it from it's state in the cluster -- must be different from `configMapOriginalState`.

oomichi replied:
> configMapFromUpdate has the value of configMap after it's updated, pulling it from it's state in the cluster -- must be different from configMapOriginalState.

Yes right. Then it is better to check not only difference from the original but also that the updated data should be the same as the data which is used for updating (`"data": "value"`)

BobyMCbobs replied:
Updating the Expect statement in 7218182f119.

oomichi reviewed (commented):
Most part is good for me, just one comment for one more check.

BobyMCbobs committed:
Update configMap data checking

BobyMCbobs commented:
/test pull-kubernetes-integration

oomichi commented:
/retest

oomichi commented:
@BobyMCbobs Thanks for updating

/lgtm
